### PR TITLE
fix: support snapshot with `expect.soft`

### LINKED ISSUE #8673: expect.soft doesnt work with toMatchSnapshot
### Describe the bug

Expect.soft doesn't seem to be respected when using toMatchSnapshot, and instead of continuing the test it terminates on failed execution.

```
    expect.soft({foo:"bar"}).toMatchObject({foo:"baz"})
    // execution continues
 
    expect.soft({ time: new Date() }).toMatchSnapshot()
    // execution doesn't continue here
```

### Reproduction

https://github.com/tijmenvanegmond/vitest-expect-soft-snapshota

### System Info

```shell
System:
    OS: macOS 26.0.1
    CPU: (8) arm64 Apple M1
    Memory: 315.45 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 22.14.0
    npm: 10.9.2
    pnpm: 10.0.0
  Browsers:
    Chrome: 141.0.7390.55
    Firefox: 143.0.4
    Safari: 26.0.1
  npmPackages:
    @vitest/ui: 3.2.4 => 3.2.4 
    vitest: 3.2.4 => 3.2.4
```

### Used Package Manager

pnpm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by iumehara**
### Description

Resolves #8673
- `toMatchSnapshot`, `matchSnapshot`, `toMatchFileSnapshot`, `toThrowErrorMatchingSnapshot` can be used with `expect.soft`. These methods use the existing `wrapAssertion` function from expect utils.
- ~`toMatchInlineSnapshot`, `toThrowErrorMatchingInlineSnapshot` could not easily support `expect.soft`. These methods now throw an error if used with `expect.soft`~ (EDIT: inline snapshots now supported)

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.
- [x] Please check [Allow edits by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to make review process faster. Note that this option is not available for repositories that are owned by Github organizations.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Review by hi-ogawa — COMMENTED** [review]
Can you create tests in `test/snapshots`? `test/core` only allows single run, so it's probably hard to poke the edge cases here (e.g. creation scenario, update scenario, non-match scenario, etc)

**Comment by iumehara** [comment]
@hi-ogawa , can you clarify what I should test in `test/snapshots`?
It seems like the edge cases (ie. creation, update) will ALWAYS pass. Is that not the case? If so, it'll never exercise the new behavior, which is a side-effect of a failing snapshot test.
Is it that you would like test coverage for assurance that expet.soft has no unexpected side-effects in the success scenario? Or is there something else that I am missing?

**Comment by hi-ogawa** [comment]
Sorry, my suggestion was a bit vague and I don't have concrete suggestion. Some ideas are:
- it looks like currently tests are relying on manipulating `snapshotState` and `result.state` from tests itself to simulate certain scenarios. this feels a bit brittle and ideally each scenario should be tested in a more realistic way like users would actually write tests.
- for example, `test/snapshots/test/skip-test.test.ts` runs same test suites multiple times with different config/snapshot to test various scenarios.

**Comment by hi-ogawa** [comment]
For example, one "brittleness" is that current tests would break with force update `pnpm test snapshot-test -u`. So this type of tests should belong to elsewhere in general.

**Comment by iumehara** [comment]
@hi-ogawa , thanks for your detailed input.
I moved the testing to the snapshots directory like you suggested.
I defined an [AlwaysFailingSnapshotSerializer](https://github.com/vitest-dev/vitest/pull/9231/commits/c480859732cb5f3c001424eade3777be9e4b1fef#diff-24f56550e1850229151abe3919809777656e462ab76c382f87e3b0863819cb42R4) in order to test the failure tracking behavior across the different scenarios.
Please let me know what you think.

**Comment by sheremet-va** [comment]
CI is failing in your tests (rolldown is fine to fail and webriverio on windows sometimes hangs, but other tests seem to come from the changes in this PR)

File: packages/vitest/src/integrations/snapshot/chai.ts [hunk 1]
```diff
@@ -119,13 +115,19 @@ export const SnapshotPlugin: ChaiPlugin = (chai, utils) => {
         ...getTestNames(test),
       })
 
+      // const isSoft = utils.flag(this, 'soft')
+      // if (isSoft) {
+      //   // assert result with soft wrapAssertion instead of recordAsyncExpect
+      //   return promise
+      // }
```
Comment (hi-ogawa) [inline packages/vitest/src/integrations/snapshot/chai.ts]: I wasn't sure why this check is needed specifically. Let me test run by commenting out. Please let me know if there's a reason for this.

**Review by hi-ogawa — COMMENTED** [review]
Hey, thanks for the PR and sorry for the long wait. I added some integration test to simulate more proper user scenario. Admittedly current `test/snapshots` suite is quite confusing, but the test like this was what I had in mind. I appreciate your effort :pray:

**Comment by hi-ogawa** [comment]
In principle, it looks possible to support soft inline snapshot, but likely requires tweaking stack probing of `__INLINE_SNAPSHOT__`. 

https://github.com/vitest-dev/vitest/blob/92ab06d6c2e6cb13885887eddb7acb716c186439/packages/snapshot/src/port/state.ts#L171-L176

Let me look into this a bit.

File: packages/vitest/src/integrations/snapshot/chai.ts (hunk 1 above)
Comment (hi-ogawa) [reply packages/vitest/src/integrations/snapshot/chai.ts]: Oh wait, I realized that this should be actually good to keep since we should still make `recordAsyncExpect` it functional for non soft case. We just didn't have a test to catch this.

EIDIT: Ah no. `wrapAssertion` always exists regardless of soft or not, so `recordAsyncExpect`'s hanging promise detection won't work (but still auto await should be working).

File: test/snapshots/README.md
Comment (hi-ogawa) [inline test/snapshots/README.md:1]: Added a quick guide for snapshot test suites. We can likely simplify `test:generate -> update` combo to unify within a single `test:integration` later.

File: packages/snapshot/src/port/state.ts [hunk 2]
```diff
@@ -168,6 +168,15 @@ export default class SnapshotState {
...
     }
 
+    // inline snapshot function can be named __INLINE_SNAPSHOT_OFFSET_<n>__
+    // to specify a custom stack offset
+    for (let i = 0; i < stacks.length; i++) {
+      const match = stacks[i].method.match(/__INLINE_SNAPSHOT_OFFSET_(\d+)__/)
+      if (match) {
+        return stacks[i + Number(match[1])] ?? null
+      }
+    }
```
Comment (hi-ogawa) [inline packages/snapshot/src/port/state.ts:178]: Added minor enhancement on `@vitest/snapshot` side to support more flexible inline snapshot stack offset.

File: packages/expect/src/utils.ts [hunk 3]
```diff
@@ -39,6 +41,12 @@ export function recordAsyncExpect(
     if (!test.promises) {
       test.promises = []
     }
+    // setup `expect.soft` handler here instead of `wrapAssertion` to avoid double tracking
+    if (isSoft) {
+      promise = promise.then(noop, (err) => {
+        handleTestError(test, err)
+      })
+    }
```
Comment (hi-ogawa) [inline packages/expect/src/utils.ts]: There are likely several ways to make `soft` and async assertion work together. Some coupling of `recordAsyncExpect` and `wrapAssertion` is imminent and I went with this approach of `recordAsyncExpect` taking over soft logic. Not too clean but code wise simplest.

**Review by hi-ogawa — DISMISSED** [review]

File: packages/expect/src/utils.ts (hunk 3 above)
Comment (hi-ogawa) [reply packages/expect/src/utils.ts]: Actually if `recordAsyncExpect` is handing `expect.soft` logic, `wrapAssertion` shouldn't be needed. Rethinking. 

EDIT: https://github.com/vitest-dev/vitest/pull/9231/commits/658a1a1811ccc35ad20ae8a45170d6828db81010

**Review by hi-ogawa — APPROVED** [review]

**Review by sheremet-va — APPROVED** [review]
